soc: nuvoton: numaker: add poweroff for m46x

cyliangtw opened the pull request:
This PR is to support sys_poweroff API on NuMaker M46x series.
About the power down mode, it could config standby or deep power down in device tree.
More, to provide system_off sample to illustrate how to enter power down by sys_poweroff API.

cyliangtw committed:
soc: nuvoton: numaker: add poweroff for m46x

Add support of sys_poweroff API on m46x series.
It could support SPD standby or DPD deep power down mode.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

fabiobaltieri commented on dts/bindings/clock/nuvoton,numaker-scc.yaml:
```diff
@@ -47,3 +47,8 @@ properties:
+  pd-mode:
+    type: int
+    description: |
+      Configure power down mode
```
Maybe expand on the description on this one, mention that this is one of the `NUMAKER_CLK_PMUCTL_PDMSEL_` macros. Also are these the same on all devices in the family? Maybe you could use an an enum instead.

bjarki-andreasen replied:
Maybe use a string instead with the enum:
```
type: string
enum:
  - "pd"
  - "llpd"
  - "fwpd"
description: |
  pd - normal power down
```

cyliangtw replied:
Thanks of your advice, it's good to mention to choose one of NUMAKER_CLK_PMUCTL_PDMSEL_XXX from numaker_xxx_clock.h
It's not the same on all devices, so keep int is better.

bjarki-andreasen replied:
I see, in this case, you should split the file `numaker_m46x_clock.h` into the specific chips you support, this ensures only valid settings for that chip are available :)

You will have:
`numaker_m461_clock.h`
`numaker_m462_clock.h`
...
where `numaker_m461_clock.h` only includes defines valid for that chip :)

cyliangtw replied:
@bjarki-andreasen ,
m46x series have the same power down modes, so current numaker_m46x_clock.h is general for m46x series.
I mentioned all devices is meaning to include other chip series, like as m2l31x series with SPD0 to 2 and DPD0 to 1 modes, I will add these modes into numaker_m2l31x_clock.h in next PR.
Appreciate your advice.

fabiobaltieri commented on soc/nuvoton/numaker/m46x/poweroff.c:
```diff
@@ -0,0 +1,35 @@
+	CLK->PMUSTS |= CLK_PMUSTS_CLRWK_Msk;
+
+	/* Select Power-down mode */
+#if DT_NODE_HAS_PROP(DT_NODELABEL(scc), pd_mode)
```
You could set a default in the property, then you would not need to check if the prop is set, the DT parser will always set it for you.

cyliangtw replied:
Yes, to use DT_PROP_OR

fabiobaltieri commented on samples/boards/numaker/system_off/src/main.c:
```diff
@@ -0,0 +1,65 @@
+#include <zephyr/sys/poweroff.h>
+
+/* Fri Jan 01 2021 13:29:50 GMT+0000 */
+#define RTC_TEST_ALARM_SET_TIME		      (1609507790)
+#define RTC_TEST_ALARM_TIME_MINUTE	      (30)
+#define RTC_TEST_ALARM_TIME_HOUR	      (13)
```
nit: no need for parenthesis on these ones

fabiobaltieri commented on samples/boards/numaker/system_off/src/main.c:
```diff
@@ -0,0 +1,65 @@
+
+static const struct device *rtc = DEVICE_DT_GET(DT_ALIAS(rtc));
+
+void set_alarm_10s(void)
```
`static`

cyliangtw commented on dts/bindings/clock/nuvoton,numaker-scc.yaml:
```diff
@@ -47,3 +47,8 @@ properties:
     description: |
       Configure core clock (HCLK)
+
+  pd-mode:
```
yes, powerdown-mode is more clear.

cyliangtw commented:
@bjarki-andreasen ,
About type of pd-mode, our another chip M2L31 has SPD0~2 & DPD0~1, so different chip has different enum. Thus, I prefer to let user refer to numaker_xxx_clock.h .

kartben commented on samples/boards/numaker/system_off/src/main.c:
```diff
@@ -0,0 +1,65 @@
+/*
```
Why is this introduced as a board specific sample? Did I miss something preventing this one to be generic?

cyliangtw replied:
it's similar to \zephyr\samples\boards\stm32\power_mgmt\serial_wakeup\src.
And current sys_poweroff samples are in boards:
```
$ find -name "*.c" | xargs grep --color sys_poweroff
./boards/esp32/deep_sleep/src/main.c:   sys_poweroff();
./boards/mimxrt595_evk/system_off/src/main.c:   sys_poweroff();
./boards/nrf/system_off/src/main.c:     sys_poweroff();
./boards/numaker/system_off/src/main.c: sys_poweroff();
./boards/stm32/power_mgmt/standby_shutdown/src/main.c:          sys_poweroff();
./boards/stm32/power_mgmt/stm32wb_ble/src/main.c:       sys_poweroff();
./boards/stm32/power_mgmt/wkup_pins/src/main.c: sys_poweroff();
./boards/ti/cc13x2_cc26x2/system_off/src/main.c:        sys_poweroff();

```

bjarki-andreasen replied:
Usually this is because enabling the wakeup source is board/soc specific (we still don't have a generic way of doing this in-tree), however, this PR enables the RTC wakeup source from within the sys_poweroff implementation, so it can be generic :)

cyliangtw replied:
Thanks of your advice.
After review, there are some pins relative to board and decide to disable these specific pins to save power in power down mode.

kartben replied:
ok, I guess it's not generic anymore :)

cyliangtw committed:
samples: boards: Add numaker system_off sample

Increase numaker board and add system_off sample
to illustrate power down by sys_poweroff API.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>
